pkg/models/devops/devops.go

```diff
@@ -1009,3 +1013,133 @@ func parseBody(body io.Reader) (newReqBody io.ReadCloser) {
 	}
 	return rc
 }
+
+func (d devopsOperator) BuildPipelineParameters(projectName string, pipelineName string, query url.Values) ([]devopsv1alpha3.ParameterDefinition, error) {
+	projectObj, err := d.ksclient.DevopsV1alpha3().DevOpsProjects().Get(d.context, projectName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	pipelineObj, err := d.ksclient.DevopsV1alpha3().Pipelines(projectObj.Status.AdminNamespace).Get(d.context, pipelineName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	var allParams []devopsv1alpha3.ParameterDefinition
+	if pipelineObj.Spec.Pipeline == nil {
+		// if pipeline type is multi-branch, direct return a empty slice
+		return allParams, nil
+	}
+	externalParams := d.buildParametersRef(pipelineObj.Spec.Pipeline.ParametersFrom, query)
+	allParams = append(allParams, pipelineObj.Spec.Pipeline.Parameters...)
+	allParams = append(allParams, externalParams...)
+	// remove items with duplicated name
+	ret := mergeParameters(allParams)
+
+	return ret, nil
+}
+
+func (d devopsOperator) buildParametersRef(refs []devopsv1alpha3.ParameterReference, query url.Values) []devopsv1alpha3.ParameterDefinition {
+	var ret []devopsv1alpha3.ParameterDefinition
+	for _, param := range refs {
+		var params []devopsv1alpha3.ParameterDefinition
+		var err error
+		if param.Kind == "ConfigMap" {
+			if param.Mode == devopsv1alpha3.PARAM_REF_MODE_CONFIG {
+				params, err = buildParamFromConfigMapData(d.k8sclient, param.Name, param.ValuesKey)
+				if err != nil {
+					klog.Errorf("buildParamFromConfigMapData error:[%v]", err)
+					continue
+				}
+
+			} else if param.Mode == devopsv1alpha3.PARAM_REF_MODE_RESTFUL {
+				params, err = buildParamFromRESTfull(d.k8sclient, param.Name, query)
+				if err != nil {
+					klog.Errorf("buildParamFromRESTfull error:[%v]", err)
+					continue
+				}
+			}
+		}
+		ret = append(ret, params...)
+	}
+	return ret
+}
+
+func mergeParameters(params []devopsv1alpha3.ParameterDefinition) []devopsv1alpha3.ParameterDefinition {
+	var ret []devopsv1alpha3.ParameterDefinition
+	tmpMap := make(map[string]byte)
+	for i := 0; i < len(params); i++ {
+		idx := len(params) - 1 - i
+		cfgName := params[idx].Name
+		if _, ok := tmpMap[cfgName]; !ok {
+			ret = append(ret, params[idx])
+			tmpMap[cfgName] = 1
+		}
+	}
+	return ret
+}
+
+func checkParametersCMName(fullName string) (namespace string, name string, err error) {
+	nameInfo := strings.Split(fullName, "/")
+	if len(nameInfo) != 2 {
+		return "", "", fmt.Errorf("invalid name [%s]", fullName)
+	}
+	return nameInfo[0], nameInfo[1], nil
+}
+
+func buildParamFromConfigMapData(k8sclient kubernetes.Interface, name, valuesKey string) ([]devopsv1alpha3.ParameterDefinition, error) {
+	var ret []devopsv1alpha3.ParameterDefinition
+	cmNS, cmName, err := checkParametersCMName(name)
```

Comment:
As we can see the reference type is `TypedLocalObjectReference`. It's not good to let it be able to use any namespaces.
Allowing it can access ConfigMap from any namespaces might have potential security issues. Normally, a resource can only access a ConfigMap from the current namespace. For example, a deploy can only mount a ConfigMap from the current namespace.